logstash_async/transport.py

```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+    :type username: str
+    :param password: Password for basic authorization (Default: "")
+    :type password: str
+    """
+
+    def __init__(
+            self,
+            host,
+            port,
+            timeout=2.0,
+            ssl_enable=True,
+            ssl_verify=True,
+            **kwargs
+    ):
+        super().__init__(host, port, timeout, ssl_enable, ssl_verify)
+        self.username = kwargs.get('username', None)
+        self.password = kwargs.get('password', None)
+        self.__session = None
+        self.__max_attempts = 3
+
+    @property
+    def url(self):
+        """The URL of the logstash pipeline based on the hostname, the port and
+        the TLS usage.
+
+        :return: The URL of the logstash pipeline
+        :rtype: str
+        """
+        protocol = 'http'
+        if self.ssl_enable:
+            protocol = 'https'
+        return '{}://{}:{}'.format(protocol, self.host, self.port)
+
+    def encode(self, events):
+        """Decodes a list of events
+        :param events: A list of events
+        :type events: list
+        :return: A list of decoded events
+        :rtype: str
+        """
+        return [json.loads(event) for event in events]
+
+    def __auth(self):
+        """The authentication method for the logstash pipeline. If the username
+        or the password is not set correctly it will return None.
+
+        :return: A HTTP basic auth object or None
+        :rtype: HTTPBasicAuth
+        """
+        if self.username is None or self.password is None:
+            return None
+        return HTTPBasicAuth(self.username, self.password)
+
+    def close(self):
+        """The HTTP connection does not need to be closed because it's
+        stateless.
+        """
+        if self.__session is not None:
+            self.__session.close()
+
+    def __backoff(self, attempt, cap=3000, base=10):
+        return random.randrange(0, min(cap, base * 2 ** attempt))
+
+    def send(self, events, **kwargs):
+        """Send events to the logstash pipeline
+
+        :param events: A list of events
+        :type events: list
+        :param use_logging: Not used!
+        :type use_logging: bool
+        """
+        headers = {'Content-Type': 'application/json'}
+        self.__session = requests.Session()
+        attempt = 0
+        while attempt < self.__max_attempts:
+            response = requests.post(
+                self.url,
+                headers=headers,
+                json=self.encode(events),
+                verify=self.ssl_verify,
+                auth=self.__auth())
+            status_code = response.status_code
+            if status_code == 200:
+                break
+            if status_code == 429:
+                sleep(self.__backoff(attempt))
+                attempt += 1
+            else:
+                self.close()
+                error_msg = 'Logstash respond with error {}'.format(status_code)
```

Comment:
`responded` might be better as we got the response already?
More importantly, we should also include the text response from Logstash. I assume this will help more than just reading "there was a 500" :).

Reply:
I will check how the plugin handles errors in general and try to add the message here.

Reply:
Maybe it's enough to simply add `requests.text`.

Reply:
I use now `response.reason`